flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogHandleReaderWithCache.java

```diff
@@ -0,0 +1,184 @@
+                        .map(path -> new File(path, CACHE_FILE_SUB_DIR))
+                        .toArray(File[]::new);
+
+        this.cacheCleanScheduler =
+                SchedulerFactory.create(1, "ChangelogCacheFileCleanScheduler", LOG);
+        // TODO: 2022/5/31 consider adding a new options for cache idle
+        this.cacheIdleMillis = config.get(PERIODIC_MATERIALIZATION_INTERVAL).toMillis();
+        this.cacheFileCreator = RefCountedTmpFileCreator.inDirectories(tempFiles);
+    }
+
+    @Override
+    public DataInputStream apply(StreamStateHandle handle, Long offset) throws IOException {
+        if (!(handle instanceof FileStateHandle)) {
+            return wrapAndSeek(handle.openInputStream(), offset);
+        }
+
+        FileStateHandle fileHandle = (FileStateHandle) handle;
+        DataInputStream input;
+
+        if (fileHandle.getFilePath().getFileSystem().isDistributedFS()) {
+
+            Path dfsPath = fileHandle.getFilePath();
+
+            final RefCountedBufferingFileStream refCountedFileStream =
+                    cache.computeIfAbsent(
+                            dfsPath,
+                            key -> {
+                                RefCountedBufferingFileStream fileStream = null;
+                                FSDataInputStream handleInputStream = null;
+
+                                try {
+                                    fileStream =
+                                            RefCountedBufferingFileStream.openNew(cacheFileCreator);
+                                    handleInputStream = handle.openInputStream();
+                                    IOUtils.copyBytes(wrap(handleInputStream), fileStream);
+                                    LOG.debug(
+                                            "download and decompress dstl file : {} to cached file : {}",
+                                            fileHandle.getFilePath(),
+                                            fileStream.getInputFile().getPath());
+                                } catch (IOException e) {
+                                    ExceptionUtils.rethrow(e);
+                                } finally {
+                                    if (handleInputStream != null) {
+                                        IOUtils.closeQuietly(handleInputStream);
+                                    }
+                                }
+
+                                return fileStream;
+                            });
+
+            FileInputStream fin = null;
+            synchronized (cache) {
+                if (refCountedFileStream.getReferenceCounter() >= 1) {
+                    fin = new FileInputStream(refCountedFileStream.getInputFile());
+                    if (offset != 0) {
+                        LOG.debug("seek to {}", offset);
+                        fin.getChannel().position(offset);
+                    }
+                    refCountedFileStream.retain();
```

Comment:
I think `retain` must be called much earlier - even before adding an entry to the map (in `computeIfAbsent`). Otherwise, the entry can be invalidated while this thread is using it.

Reply:
I see `retain` is still called **after** `computeIfAbsent`. That forces to:
- add `synchronized`
- re-download the file if it was deleted in the meantime (
```
// cache file already be deleted, try again
return getRefCountedFile(fileHandle);
```
)

Are there any reasons NOT to put retain inside `downloadToCacheFile` ?